### ISSUE TITLE
Batch_size by tokens (for transformer)

### ISSUE THREAD

**Issue by srush** [OpenNMT/OpenNMT-py#465]
Hey  @helson73  cc. @JianyuZhan 

For your batch size with tokens PR, I would like the following additions:

1) Let's have clear options: maybe add `-batch_size_tokens [num]` and `-batch_size_type [sents, tokens]` defaults to sents for now. This will be nice to have for many of the models. 

2) There is a nice option in iterator that you can use for batch_size selection called `batch_size_fn`. I would use that instead of writing your own. 

https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L50

3)  The learning rate schedule we have in optim might be wrong.  Feel free to make it closer to tf. 

https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Optim.py#L61

Thanks so much for these changes. I've really wanted to do this but haven't had the time.

**Comment by helson73** [comment]
Thanks, I will take a look next week!

**Comment by srush** [comment]
Think this is all implemented and working.
